### PR TITLE
openpyxl: `Cell | MergedCell` annotations

### PR DESCRIPTION
**Pull request by Avasam**
Extracted and completed from https://github.com/python/typeshed/pull/9511

This basically comes down to typing `Worksheet._cells` then working up the tree of references.
Then I checked all existing `Cell` annotations as a sanity check.

This does result in a few extra `Cell | MergedCell` return types, which may negatively affect some users since no Cell-like type inherits from `Cell`. This is the same issue as https://github.com/python/typeshed/pull/11718 and could be solved in the same way.

Note that this does *not* address https://github.com/python/typeshed/issues/9940

### CODE REVIEW

File: stubs/openpyxl/openpyxl/worksheet/worksheet.pyi [hunk 1]
```diff
@@ -51,6 +51,7 @@ class Worksheet(_WorkbookChild):
     ORIENTATION_PORTRAIT: Final = "portrait"
     ORIENTATION_LANDSCAPE: Final = "landscape"
 
+    _cells: dict[tuple[int, int], _CellOrMergedCell] # private but very useful to understand typing
```
Comment (Avasam) [inline stubs/openpyxl/openpyxl/worksheet/worksheet.pyi]: I could just comment this out before merging, Users shouldn't be using this, but it's a useful note.

**Review by srittau — APPROVED** [review]
Thanks!

File: stubs/openpyxl/openpyxl/worksheet/worksheet.pyi [hunk 2]
```diff
@@ -132,11 +135,13 @@ class Worksheet(_WorkbookChild):
         min_col: int | None = None,
         max_col: int | None = None,
         values_only: Literal[False] = False,
-    ) -> Generator[tuple[Cell, ...], None, None]: ...
+    ) -> Generator[tuple[_CellOrMergedCell, ...], None, None]: ...
     @overload
     def iter_rows(
         self, min_row: int | None, max_row: int | None, min_col: int | None, max_col: int | None, values_only: bool
-    ) -> Generator[tuple[Cell, ...], None, None] | Generator[tuple[str | float | datetime | None, ...], None, None]: ...
+    ) -> (
```
Comment (mutricyl) [inline stubs/openpyxl/openpyxl/worksheet/worksheet.pyi:142]: Is this change related to issue #13151 ?
